scripts/gen.py

```diff
@@ -0,0 +1,429 @@
+#!/usr/bin/python
+
+from __future__ import print_function
+
+import argparse
+import collections
+import lark
+import os
+import re
+import sys
+
+FuncGen = collections.namedtuple(
+    'FuncGen', 'tree, xtree, func, xfunc, code, sig, cppsig, funsig, mapsig')
+
+_GRAMMAR = r"""
+    start: type fnname "(" params ")"
+    type: CONST? core_type refspec?
+    fnname: CNAME
+    refspec: REF
+           | PTR
+    core_type: template
+        | TNAME
+    template: TNAME "<" typelist ">"
+    typelist: type
+            | type "," typelist
+    REF: "&"
+    PTR: "*"
+    CONST: "const"
+    TNAME: /[a-zA-Z0-9_:]+/
+    params: param
+          | param "," params
+    param: type param_name
+    param_name: CNAME
+
+    %import common.CNAME -> CNAME
+    %import common.WS
+    %ignore WS
+    """
+
+_PARSER = lark.Lark(_GRAMMAR, parser='lalr', propagate_positions=True)
+_XPARSER = lark.Lark(
+    _GRAMMAR, parser='lalr', propagate_positions=True, keep_all_tokens=True)
+
+_FN_BLACKLIST = set([
+    # ATEN functions
+    'toBackend',
+    'toScalarType',
+    'copy',
+    'copy_',
+    'backward',
+    'tensorFromBlob',
+    'tensorWithAllocator',
+    'storageFromBlob',
+    'storageWithAllocator',
+    'unsafeStorageFromTH',
+    'unsafeTensorFromTH',
+    # XLA/TPU functions
+])
+_FN_BLACKLIST_REGEX = [
+    r'.*cudnn',
+]
+
+
+def first_match(t):
+  if isinstance(t, lark.lexer.Token):
+    return t.column - 1
+  assert isinstance(t, lark.tree.Tree)
+  return first_match(t.children[0])
+
+
+def last_match(t):
+  if isinstance(t, lark.lexer.Token):
+    return t.end_column - 1
+  assert isinstance(t, lark.tree.Tree)
+  return last_match(t.children[-1])
+
+
+class StringEmit(object):
+
+  def __init__(self, sref):
+    self.sref = sref
+    self.sval = ''
+    self.pos = -1
+
+  def __repr__(self):
+    return self.sval
+
+  def advance(self, t):
+    start = first_match(t)
+    end = last_match(t)
+    pos = self.pos if self.pos >= 0 else start
+    self.sval += self.sref[pos:end]
+    self.pos = end
+
+  def skip(self, t):
+    self.pos = last_match(t) if self.pos >= 0 else -1
+
+  def append(self, s):
+    self.sval += s
+    self.pos = -1
+
+
+def list_get(l, n):
+  return l[n] if n < len(l) else None
+
+
+def is_blacklisted_fn(fname):
+  if fname in _FN_BLACKLIST:
+    return True
+  for frx in _FN_BLACKLIST_REGEX:
+    if re.match(frx, fname):
+      return True
+  return False
+
+
+def for_every_token(t, fn):
+  if isinstance(t, lark.lexer.Token):
+    fn(t)
+  else:
+    assert isinstance(t, lark.tree.Tree)
+    for c in t.children:
+      for_every_token(c, fn)
+
+
+def emit_string(t, emit, emit_fn):
+  status = emit_fn(t)
+  if status > 0:
```

Comment:
When is this branch taken? All `emit_fn` I see return either 0 or -1.

Reply:
It's a feature that allows to whitelist a whole Tree while enumerating it for emission.